src/QsCompiler/SyntaxProcessor/CapabilityInference/TypeAnalyzer.fs

```diff
@@ -0,0 +1,210 @@
+    ty.ExtractAll(shallowFindScenario context construct >> Option.toList >> Seq.ofList)
+
+let scenarioCapability scenario =
+    let classical =
+        match scenario with
+        | UseBigInt -> ClassicalCapability.full
+        | StringNotArgumentToMessage -> ClassicalCapability.full
+        | ConditionalOrMutable ty ->
+            match ty.Resolution with
+            | Bool
+            | Int -> ClassicalCapability.integral
+            | _ -> ClassicalCapability.full
+        | EntryPointParam ty ->
+            match ty.Resolution with
+            | TupleType _ -> ClassicalCapability.empty
+            | Bool
+            | Int -> ClassicalCapability.integral
+            | _ -> ClassicalCapability.full
+        | EntryPointReturn ty ->
+            match ty.Resolution with
+            | UnitType
+            | Result
+            | TupleType _
+            | ArrayType _ -> ClassicalCapability.empty
+            | Bool
+            | Int -> ClassicalCapability.integral
+            | _ -> ClassicalCapability.full
+
+    TargetCapability.withClassical classical TargetCapability.bottom
+
+let shallowTypeName (ty: ResolvedType) =
+    match ty.Resolution with
+    | ArrayType _ -> "array"
+    | TupleType _ -> "tuple"
+    | QsTypeKind.Operation _ -> "operation"
+    | QsTypeKind.Function _ -> "function"
+    | _ -> SyntaxTreeToQsharp.Default.ToCode ty
+
+let describeScenario =
+    function
+    | UseBigInt -> "BigInt"
+    | StringNotArgumentToMessage -> "string that is not an argument to Message"
+    | ConditionalOrMutable ty -> "conditional expression or mutable variable of type " + shallowTypeName ty
+    | EntryPointParam ty -> "entry point parameter of type " + shallowTypeName ty
+    | EntryPointReturn ty -> "entry point return value of type " + shallowTypeName ty
+
+let createPattern context construct range (ty: ResolvedType) =
+    let scenarios = findScenarios context construct ty |> Seq.toList
+
+    let capability =
+        Seq.map scenarioCapability scenarios |> Seq.fold TargetCapability.merge TargetCapability.bottom
+
+    if capability = TargetCapability.bottom then
+        None
+    else
+        let diagnose (target: Target) =
+            if TargetCapability.subsumes target.Capability capability then
+                None
+            else
+                let unsupported =
+                    Seq.filter (scenarioCapability >> TargetCapability.subsumes target.Capability >> not) scenarios
+
+                let description = Seq.map describeScenario unsupported |> String.concat ", "
+                let args = [ target.Name; description ]
+                let range = QsNullable.defaultValue Range.Zero range
+                QsCompilerDiagnostic.Error(ErrorCode.UnsupportedClassicalCapability, args) range |> Some
+
+        Some
+            {
+                Capability = capability
+                Diagnose = diagnose
+                Properties = ()
+            }
+
+let rec flattenTuple =
```

Comment:
This one keeps being redefined on every other PR. The SyntaxGenerator contains a method for this, and I know there has been a desire to have a C# extension to flatten argument tuples. Maybe it would be a good time to add that?

Reply:
I see, I didn't know about QsTuple.Items. I removed `flattenTuple`. I don't think I need to add anything new right now.